shared/src/hints.rs

```diff
@@ -0,0 +1,44 @@
+//! Compiler hints, to improve the performance of certain operations.
+
+/// A function that is seldom called.
+#[inline]
+#[cold]
+pub fn cold() {}
+
+/// A likely path to be taken in an if-expression.
+///
+/// # Example
+///
+/// ```ignore
+/// if likely(frequent_condition()) {
+///     // most common path to take
+/// } else {
+///     // ...
+/// }
+/// ```
+#[inline]
+pub fn likely(b: bool) -> bool {
+    if !b {
+        cold()
+    }
+    b
+}
+
+/// An unlikely path to be taken in an if-expression.
+///
+/// # Example
+///
+/// ```ignore
+/// if unlikely(rare_condition()) {
+///     // ...
+/// } else {
+///     // most common path to take
+/// }
+/// ```
+#[inline]
+pub fn unlikely(b: bool) -> bool {
+    if b {
+        cold()
+    }
+    b
+}
```

Comment:
Should go towards `main`, idk much about this. I'd be slightly hesitant to have optimizations like this at this stage (we should be able to do this sort of thing by profiling anyway - https://doc.rust-lang.org/rustc/profile-guided-optimization.html ?).

Reply:
I admit I was having way too much fun when I added this module haha